File: language-samples/hello.go
/**
 * Go Advanced Sample - Demonstrating Most Powerful Features
 * BarbrickDesign - Complete Language Portfolio
 *
 * Features demonstrated:
 * - Goroutines & Concurrency
 * - Channels & Select
 * - Interfaces & Embedding
 * - Defer, Panic, Recover
 * - Context & Cancellation
 * - Generics (Go 1.18+)
 * - Error Handling Patterns
 * - Struct Tags & Reflection
 */

package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"reflect"
	"sync"
	"time"
)

// ============ Interfaces ============

type Processor interface {
	Process(data interface{}) (interface{}, error)
	Validate() bool
}

type Logger interface {
	Info(msg string)
	Error(msg string, err error)
}

// ============ Error Types ============

var (
	ErrNotFound      = errors.New("not found")
	ErrInvalidInput  = errors.New("invalid input")
	ErrTimeout       = errors.New("operation timeout")
)

type ValidationError struct {
	Field   string
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("validation error on %s: %s", e.Field, e.Message)
}

// ============ Generic Functions (Go 1.18+) ============

func Map[T any, U any](slice []T, fn func(T) U) []U {
	result := make([]U, len(slice))
	for i, v := range slice {
		result[i] = fn(v)
	}
	return result
}

func Filter[T any](slice []T, fn func(T) bool) []T {
	result := make([]T, 0)
	for _, v := range slice {
		if fn(v) {
			result = append(result, v)
		}
	}
	return result
}

func Reduce[T any, U any](slice []T, initial U, fn func(U, T) U) U {
	result := initial
	for _, v := range slice {
		result = fn(result, v)
	}
	return result
}

// ============ Generic Types ============

type Cache[K comparable, V any] struct {
	mu    sync.RWMutex
	items map[K]V
}

func NewCache[K comparable, V any]() *Cache[K, V] {
	return &Cache[K, V]{
		items: make(map[K]V),
	}
}

func (c *Cache[K, V]) Set(key K, value V) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.items[key] = value
}

func (c *Cache[K, V]) Get(key K) (V, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	val, exists := c.items[key]
	return val, exists
}

// ============ Structs with Tags ============

type User struct {
	ID        string    `json:"id" validate:"required"`
	Name      string    `json:"name" validate:"required,min=2"`
	Email     string    `json:"email" validate:"required,email"`
	Age       int       `json:"age" validate:"min=0,max=150"`
	CreatedAt time.Time `json:"created_at"`
}

func (u *User) Validate() error {
	if u.Name == "" {
		return &ValidationError{Field: "Name", Message: "cannot be empty"}
	}
	if u.Age < 0 || u.Age > 150 {
		return &ValidationError{Field: "Age", Message: "must be between 0 and 150"}
	}
	return nil
}

// ============ Interface Implementation ============

type UserProcessor struct {
	logger Logger
}

func (p *UserProcessor) Process(data interface{}) (interface{}, error) {
	user, ok := data.(*User)
	if !ok {
		return nil, ErrInvalidInput
	}
	
	if err := user.Validate(); err != nil {
		return nil, err
	}
	
	p.logger.Info(fmt.Sprintf("Processing user: %s", user.Name))
	return user, nil
}

func (p *UserProcessor) Validate() bool {
	return p.logger != nil
}

// ============ Logger Implementation ============

type ConsoleLogger struct{}

func (l *ConsoleLogger) Info(msg string) {
	fmt.Printf("ℹ️  %s\n", msg)
}

func (l *ConsoleLogger) Error(msg string, err error) {
	fmt.Printf("❌ %s: %v\n", msg, err)
}

// ============ Concurrency Patterns ============

// Worker Pool Pattern
func workerPool(ctx context.Context, jobs <-chan int, results chan<- int, workerCount int) {
	var wg sync.WaitGroup
	
	for i := 0; i < workerCount; i++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			for {
				select {
				case <-ctx.Done():
					return
				case job, ok := <-jobs:
					if !ok {
						return
					}
					// Simulate work
					time.Sleep(50 * time.Millisecond)
					results <- job * 2
				}
			}
		}(i)
	}
	
	wg.Wait()
	close(results)
}

// Fan-out, Fan-in Pattern
func fanOut(input <-chan int, workers int) []<-chan int {
	channels := make([]<-chan int, workers)
	for i := 0; i < workers; i++ {
		channels[i] = worker(input)
	}
	return channels
}

func worker(input <-chan int) <-chan int {
	output := make(chan int)
	go func() {
		defer close(output)
		for val := range input {
			output <- val * 2
		}
	}()
	return output
}

func fanIn(channels ...<-chan int) <-chan int {
	output := make(chan int)
	var wg sync.WaitGroup
	
	for _, ch := range channels {
		wg.Add(1)
		go func(c <-chan int) {
			defer wg.Done()
			for val := range c {
				output <- val
			}
		}(ch)
	}
	
	go func() {
		wg.Wait()
		close(output)
	}()
	
	return output
}

// ============ Pipeline Pattern ============

type Pipeline struct {
	stages []func(interface{}) (interface{}, error)
}

func NewPipeline() *Pipeline {
	return &Pipeline{stages: make([]func(interface{}) (interface{}, error), 0)}
}

func (p *Pipeline) Add(stage func(interface{}) (interface{}, error)) *Pipeline {
	p.stages = append(p.stages, stage)
	return p
}

func (p *Pipeline) Execute(input interface{}) (interface{}, error) {
	result := input
	for _, stage := range p.stages {
		var err error
		result, err = stage(result)
		if err != nil {
			return nil, err
		}
	}
	return result, nil
}

// ============ Context & Cancellation ============

func withTimeout(ctx context.Context, duration time.Duration) {
	ctx, cancel := context.WithTimeout(ctx, duration)
	defer cancel()
	
	done := make(chan bool)
	
	go func() {
		time.Sleep(200 * time.Millisecond)
		done <- true
	}()
	
	select {
	case <-ctx.Done():
		fmt.Println("⏱️  Context timeout:", ctx.Err())
	case <-done:
		fmt.Println("✓ Operation completed in time")
	}
}

// ============ Defer, Panic, Recover ============

func safeDivide(a, b int) (result int, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("panic recovered: %v", r)
		}
	}()
	
	if b == 0 {
		panic("division by zero")
	}
	
	result = a / b
	return
}

// ============ Reflection ============

func inspectStruct(obj interface{}) {
	val := reflect.ValueOf(obj)
	typ := reflect.TypeOf(obj)
	
	fmt.Printf("\nType: %s\n", typ.Name())
	
	for i := 0; i < val.NumField(); i++ {
		field := typ.Field(i)
		value := val.Field(i)
		tag := field.Tag.Get("json")
		fmt.Printf("  %s (json: %s): %v\n", field.Name, tag, value.Interface())
	}
}

// ============ Channels & Select ============

func multiplexer() {
	ch1 := make(chan string)
	ch2 := make(chan string)
	done := make(chan bool)
	
	go func() {
		time.Sleep(100 * time.Millisecond)
		ch1 <- "Message from channel 1"
	}()
	
	go func() {
		time.Sleep(150 * time.Millisecond)
		ch2 <- "Message from channel 2"
	}()
	
	go func() {
		time.Sleep(300 * time.Millisecond)
		done <- true
	}()
	
	for {
		select {
		case msg1 := <-ch1:
			fmt.Println("CH1:", msg1)
		case msg2 := <-ch2:
			fmt.Println("CH2:", msg2)
		case <-done:
			fmt.Println("Done!")
			return
		case <-time.After(500 * time.Millisecond):
			fmt.Println("Timeout!")
			return
		}
	}
}

// ============ Main Function ============

func main() {
	fmt.Println("🚀 GO ADVANCED FEATURES SHOWCASE\n")
	
	logger := &ConsoleLogger{}
	
	// 1. Generics
	logger.Info("=== Generics ===")
	numbers := []int{1, 2, 3, 4, 5}
	squared := Map(numbers, func(n int) int { return n * n })
	fmt.Printf("Squared: %v\n", squared)
	
	evens := Filter(numbers, func(n int) bool { return n%2 == 0 })
	fmt.Printf("Evens: %v\n", evens)
	
	sum := Reduce(numbers, 0, func(acc, n int) int { return acc + n })
	fmt.Printf("Sum: %d\n", sum)
	
	// 2. Generic Cache
	logger.Info("\n=== Generic Cache ===")
	cache := NewCache[string, *User]()
	user := &User{
		ID:        "u1",
		Name:      "Alice",
		Email:     "alice@example.com",
		Age:       28,
		CreatedAt: time.Now(),
	}
	cache.Set("u1", user)
	
	if cached, exists := cache.Get("u1"); exists {
		fmt.Printf("Cached user: %s\n", cached.Name)
	}
	
	// 3. Interface Implementation
	logger.Info("\n=== Interface Implementation ===")
	processor := &UserProcessor{logger: logger}
	processed, err := processor.Process(user)
	if err != nil {
		logger.Error("Processing failed", err)
	} else {
		fmt.Printf("Processed: %v\n", processed.(*User).Name)
	}
	
	// 4. Struct Tags & JSON
	logger.Info("\n=== Struct Tags & JSON ===")
	jsonData, _ := json.MarshalIndent(user, "", "  ")
	fmt.Printf("JSON:\n%s\n", jsonData)
	
	// 5. Reflection
	logger.Info("\n=== Reflection ===")
	inspectStruct(*user)
	
	// 6. Defer, Panic, Recover
	logger.Info("\n=== Defer, Panic, Recover ===")
	result, err := safeDivide(10, 2)
	fmt.Printf("10 / 2 = %d\n", result)
	
	result, err = safeDivide(10, 0)
	if err != nil {
		logger.Error("Division failed", err)
	}
	
	// 7. Goroutines & Worker Pool
	logger.Info("\n=== Worker Pool Pattern ===")
	ctx := context.Background()
	jobs := make(chan int, 10)
	results := make(chan int, 10)
	
	go workerPool(ctx, jobs, results, 3)
	
	for i := 1; i <= 5; i++ {
		jobs <- i
	}
	close(jobs)
	
	for result := range results {
		fmt.Printf("Result: %d\n", result)
	}
	
	// 8. Context & Timeout
	logger.Info("\n=== Context & Timeout ===")
	withTimeout(context.Background(), 100*time.Millisecond)
	withTimeout(context.Background(), 300*time.Millisecond)
	
	// 9. Channels & Select
	logger.Info("\n=== Multiplexer (Select) ===")
	multiplexer()
	
	// 10. Pipeline Pattern
	logger.Info("\n=== Pipeline Pattern ===")
	pipeline := NewPipeline().
		Add(func(data interface{}) (interface{}, error) {
			n := data.(int)
			return n + 10, nil
		}).
		Add(func(data interface{}) (interface{}, error) {
			n := data.(int)
			return n * 2, nil
		}).
		Add(func(data interface{}) (interface{}, error) {
			n := data.(int)
			return n * n, nil
		})
	
	pipeResult, _ := pipeline.Execute(5)
	fmt.Printf("Pipeline result: %v\n", pipeResult) // ((5+10)*2)^2 = 900
	
	// 11. Fan-out, Fan-in
	logger.Info("\n=== Fan-out, Fan-in Pattern ===")
	input := make(chan int)
	go func() {
		for i := 1; i <= 5; i++ {
			input <- i
		}
		close(input)
	}()
	
	workers := fanOut(input, 3)
	output := fanIn(workers...)
	
	for val := range output {
		fmt.Printf("Fan result: %d\n", val)
	}
	
	logger.Info("\n✅ All advanced Go features demonstrated!")
}

File: language-samples/hello.rs
/**
 * Rust Advanced Sample - Demonstrating Most Powerful Features
 * BarbrickDesign - Complete Language Portfolio
 * 
 * Features demonstrated:
 * - Ownership, Borrowing, and Lifetimes
 * - Traits and Generics
 * - Error Handling with Result/Option
 * - Pattern Matching
 * - Smart Pointers (Box, Rc, Arc, RefCell)
 * - Iterators and Closures
 * - Concurrency (threads, channels)
 * - Macros
 * - Zero-cost Abstractions
 */

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::thread;
use std::sync::mpsc;
use std::cell::RefCell;
use std::rc::Rc;
use std::fmt;

// ============ Custom Error Types ============

#[derive(Debug, Clone)]
enum DataError {
    NotFound(String),
    ValidationError(String),
    ConcurrencyError,
}

impl fmt::Display for DataError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            DataError::NotFound(msg) => write!(f, "Not Found: {}", msg),
            DataError::ValidationError(msg) => write!(f, "Validation Error: {}", msg),
            DataError::ConcurrencyError => write!(f, "Concurrency Error"),
        }
    }
}

type Result<T> = std::result::Result<T, DataError>;

// ============ Traits ============

trait Processable {
    fn process(&self) -> String;
    fn validate(&self) -> bool;
}

trait Cacheable: Clone {
    fn cache_key(&self) -> String;
}

// ============ Generic Structs with Lifetimes ============

#[derive(Debug, Clone)]
struct Entity<'a, T: Clone> {
    id: &'a str,
    data: T,
    metadata: HashMap<String, String>,
}

impl<'a, T: Clone> Entity<'a, T> {
    fn new(id: &'a str, data: T) -> Self {
        Self {
            id,
            data,
            metadata: HashMap::new(),
        }
    }
    
    fn add_metadata(&mut self, key: String, value: String) {
        self.metadata.insert(key, value);
    }
}

// ============ Advanced Pattern Matching ============

#[derive(Debug)]
enum Message {
    Text(String),
    Image { url: String, width: u32, height: u32 },
    Video { url: String, duration: u32 },
    System(SystemMessage),
}

#[derive(Debug)]
enum SystemMessage {
    Connect,
    Disconnect,
    Ping,
}

fn handle_message(msg: Message) -> String {
    match msg {
        Message::Text(content) => format!("Text: {}", content),
        Message::Image { url, width, height } => {
            format!("Image: {} ({}x{})", url, width, height)
        }
        Message::Video { url, duration } => {
            format!("Video: {} ({}s)", url, duration)
        }
        Message::System(SystemMessage::Connect) => "System: Connected".to_string(),
        Message::System(SystemMessage::Disconnect) => "System: Disconnected".to_string(),
        Message::System(SystemMessage::Ping) => "System: Pong".to_string(),
    }
}

// ============ Smart Pointers ============

struct Node {
    value: i32,
    children: Vec<Rc<RefCell<Node>>>,
}

impl Node {
    fn new(value: i32) -> Rc<RefCell<Self>> {
        Rc::new(RefCell::new(Node {
            value,
            children: Vec::new(),
        }))
    }
    
    fn add_child(&mut self, child: Rc<RefCell<Node>>) {
        self.children.push(child);
    }
}

// ============ Iterator Adaptors ============

struct Fibonacci {
    current: u64,
    next: u64,
}

impl Fibonacci {
    fn new() -> Self {
        Fibonacci { current: 0, next: 1 }
    }
}

impl Iterator for Fibonacci {
    type Item = u64;
    
    fn next(&mut self) -> Option<Self::Item> {
        let current = self.current;
        self.current = self.next;
        self.next = current + self.next;
        Some(current)
    }
}

// ============ Generic Cache with Traits ============

struct Cache<T: Cacheable> {
    data: HashMap<String, T>,
    max_size: usize,
}

impl<T: Cacheable> Cache<T> {
    fn new(max_size: usize) -> Self {
        Cache {
            data: HashMap::new(),
            max_size,
        }
    }
    
    fn insert(&mut self, item: T) {
        if self.data.len() >= self.max_size {
            if let Some(key) = self.data.keys().next().cloned() {
                self.data.remove(&key);
            }
        }
        self.data.insert(item.cache_key(), item);
    }
    
    fn get(&self, key: &str) -> Option<&T> {
        self.data.get(key)
    }
}

// ============ Trait Implementations ============

#[derive(Debug, Clone)]
struct User {
    id: String,
    name: String,
    email: String,
}

impl Processable for User {
    fn process(&self) -> String {
        format!("Processing user: {}", self.name)
    }
    
    fn validate(&self) -> bool {
        !self.name.is_empty() && self.email.contains('@')
    }
}

impl Cacheable for User {
    fn cache_key(&self) -> String {
        self.id.clone()
    }
}

// ============ Functional Programming ============

fn compose<A, B, C, F, G>(f: F, g: G) -> impl Fn(A) -> C
where
    F: Fn(A) -> B,
    G: Fn(B) -> C,
{
    move |x| g(f(x))
}

// ============ Error Handling ============

fn fetch_user(id: &str) -> Result<User> {
    if id.is_empty() {
        return Err(DataError::ValidationError("ID cannot be empty".to_string()));
    }
    
    if id == "404" {
        return Err(DataError::NotFound(format!("User {} not found", id)));
    }
    
    Ok(User {
        id: id.to_string(),
        name: "John Doe".to_string(),
        email: "john@example.com".to_string(),
    })
}

fn process_user_chain(id: &str) -> Result<String> {
    let user = fetch_user(id)?;
    
    if !user.validate() {
        return Err(DataError::ValidationError("Invalid user data".to_string()));
    }
    
    Ok(user.process())
}

// ============ Concurrency ============

fn concurrent_processing(items: Vec<i32>) -> Vec<i32> {
    let data = Arc::new(Mutex::new(Vec::new()));
    let mut handles = vec![];
    
    for item in items {
        let data_clone = Arc::clone(&data);
        let handle = thread::spawn(move || {
            let processed = item * 2;
            let mut data = data_clone.lock().unwrap();
            data.push(processed);
        });
        handles.push(handle);
    }
    
    for handle in handles {
        handle.join().unwrap();
    }
    
    let result = data.lock().unwrap();
    result.clone()
}

// ============ Channel Communication ============

fn producer_consumer_pattern() {
    let (tx, rx) = mpsc::channel();
    
    // Producer
    thread::spawn(move || {
        for i in 1..=5 {
            tx.send(i * 10).unwrap();
            thread::sleep(std::time::Duration::from_millis(10));
        }
    });
    
    // Consumer
    for received in rx {
        println!("Received: {}", received);
    }
}

// ============ Macros ============

macro_rules! create_function {
    ($func_name:ident, $operation:expr) => {
        fn $func_name(x: i32) -> i32 {
            $operation(x)
        }
    };
}

create_function!(double, |x| x * 2);
create_function!(triple, |x| x * 3);

macro_rules! hashmap {
    ($($key:expr => $value:expr),* $(,)?) => {
        {
            let mut map = HashMap::new();
            $(
                map.insert($key, $value);
            )*
            map
        }
    };
}

// ============ Main Function ============

fn main() {
    println!("🦀 RUST ADVANCED FEATURES SHOWCASE\n");
    
    // 1. Ownership and Borrowing
    println!("=== Ownership & Borrowing ===");
    let s1 = String::from("Rust");
    let s2 = &s1; // Borrowing
    println!("Borrowed: {}", s2);
    println!("Original still valid: {}", s1);
    
    // 2. Pattern Matching
    println!("\n=== Pattern Matching ===");
    let messages = vec![
        Message::Text("Hello".to_string()),
        Message::Image { url: "image.jpg".to_string(), width: 800, height: 600 },
        Message::System(SystemMessage::Connect),
    ];
    
    for msg in messages {
        println!("{}", handle_message(msg));
    }
    
    // 3. Error Handling with Result
    println!("\n=== Error Handling ===");
    match process_user_chain("user123") {
        Ok(result) => println!("✓ {}", result),
        Err(e) => println!("✗ Error: {}", e),
    }
    
    match process_user_chain("404") {
        Ok(result) => println!("✓ {}", result),
        Err(e) => println!("✗ Error: {}", e),
    }
    
    // 4. Iterators and Closures
    println!("\n=== Iterators ===");
    let fib_sequence: Vec<u64> = Fibonacci::new().take(10).collect();
    println!("Fibonacci: {:?}", fib_sequence);
    
    let numbers = vec![1, 2, 3, 4, 5];
    let sum: i32 = numbers.iter()
        .map(|x| x * 2)
        .filter(|x| x > &5)
        .sum();
    println!("Sum of doubled & filtered: {}", sum);
    
    // 5. Functional Composition
    println!("\n=== Functional Composition ===");
    let add_5 = |x: i32| x + 5;
    let multiply_3 = |x: i32| x * 3;
    let composed = compose(add_5, multiply_3);
    println!("compose(+5, *3)(10) = {}", composed(10)); // (10+5)*3 = 45
    
    // 6. Smart Pointers
    println!("\n=== Smart Pointers (Rc & RefCell) ===");
    let root = Node::new(1);
    let child1 = Node::new(2);
    let child2 = Node::new(3);
    
    root.borrow_mut().add_child(Rc::clone(&child1));
    root.borrow_mut().add_child(Rc::clone(&child2));
    
    println!("Root value: {}", root.borrow().value);
    println!("Children count: {}", root.borrow().children.len());
    
    // 7. Generic Cache
    println!("\n=== Generic Cache ===");
    let mut user_cache = Cache::<User>::new(100);
    let user = User {
        id: "u1".to_string(),
        name: "Alice".to_string(),
        email: "alice@example.com".to_string(),
    };
    
    user_cache.insert(user.clone());
    if let Some(cached) = user_cache.get("u1") {
        println!("Cached user: {}", cached.name);
    }
    
    // 8. Trait Bounds
    println!("\n=== Traits ===");
    println!("{}", user.process());
    println!("Valid: {}", user.validate());
    
    // 9. Macros
    println!("\n=== Macros ===");
    println!("double(5) = {}", double(5));
    println!("triple(5) = {}", triple(5));
    
    let map = hashmap! {
        "language" => "Rust",
        "year" => "2010",
        "paradigm" => "Multi-paradigm",
    };
    println!("HashMap created via macro: {:?}", map);
    
    // 10. Concurrency
    println!("\n=== Concurrency (Threads) ===");
    let items = vec![1, 2, 3, 4, 5];
    let results = concurrent_processing(items);
    println!("Concurrent results: {:?}", results);
    
    // 11. Channel Communication
    println!("\n=== Channel Communication ===");
    producer_consumer_pattern();
    
    // 12. Option Type
    println!("\n=== Option Type ===");
    let some_value = Some(42);
    let no_value: Option<i32> = None;
    
    if let Some(v) = some_value {
        println!("Value exists: {}", v);
    }
    
    let unwrapped = no_value.unwrap_or(0);
    println!("Unwrapped with default: {}", unwrapped);
    
    // 13. Lifetimes
    println!("\n=== Lifetimes ===");
    let id = "entity_1";
    let mut entity = Entity::new(id, "Some data".to_string());
    entity.add_metadata("created_by".to_string(), "system".to_string());
    println!("Entity ID: {}, Data: {}", entity.id, entity.data);
    
    println!("\n✅ All advanced Rust features demonstrated!");
}
